src/core/models.py

```diff
@@ -5,15 +5,25 @@
 """Collection of state objects for the Kafka relations, apps and units."""
 
 import logging
+from functools import cached_property
 
 from charms.data_platform_libs.v0.data_interfaces import Data, DataPeerData, DataPeerUnitData
 from charms.zookeeper.v0.client import QuorumLeaderNotFoundError, ZooKeeperManager
 from kazoo.client import AuthFailedError, NoNodeError
+from lightkube.core.exceptions import ApiError
+from lightkube.resources.core_v1 import Node, Pod
 from ops.model import Application, Relation, Unit
 from tenacity import retry, retry_if_result, stop_after_attempt, wait_fixed
 from typing_extensions import override
 
-from literals import INTERNAL_USERS, SECRETS_APP, Substrates
+from literals import (
+    INTERNAL_USERS,
+    SECRETS_APP,
+    SECURITY_PROTOCOL_PORTS,
+    AuthMechanism,
+    Substrates,
+)
+from managers.k8s import K8sManager
```

Comment:
It's a bit messy, I agree, but given that managers are stateless, I wondered if there was an issue initialising it here, as it's tied to a 'broker' unit/pod. I couldn't find a clean way of doing it otherwise. It shouldn't be circular at least.

As I saw it, a 'manager' is a generic business-logic unit that takes some arguments (i.e not `state` or `charm` directly). I don't see much issue in initialising a manager anywhere it's useful? What do you think?

The planned dependency tree was more focused on `charm` > `state` > `events`, with managers executing logic functionally stateless, was that a wrong interpretation?

Reply:
Also note, `managers/k8s.py` does not depend on anything from `core`.

Reply:
yes, I noticed that too. And in an earlier version of the comment I was suggesting to make K8sManager as part of core, since it is - to some extend - just parsing K8s control-plane information, so it does not sound outrageous. However then, there is also the "business logic" how services are created, and that seems a bit more manager stuff. Unfortunately it seems that `bootstrap_server` that is the reason why we need the k8s manager here, uses also the `business_logic` part. 

I don't have a solution straight away, but I felt it was worth pointing out, maybe among you guys, you can come up with a cleaner solution.  

> I don't see much issue in initialising a manager anywhere it's useful? What do you think?

Sort of. I would still respect the dependency chain (core) => (managers) => (events). Also there can be a dependency chain between managers, e.g. `general manager (e.g config) => more-specific-manager (e.g. auth)`, but if in core you import from manager, there is a risk of potential circularity there (maybe you can avoid that, right now, but it may bite back later). So more-specific manager can import from other managers, and events can import from all of the others. But core cannot import events. That would be weird, wouldn't it?